AssertionError with BinaryDyBM

Hello,
I'm using the BinaryDyBM model to learn a pattern, where the output sequence is different from the input sequence and I get an assertion error.

I'm initializing the model with the correct in_dim and out_dim, below follows the traceback.

**Traceback**
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-15-e7a1537fc04a> in <module>
      7 
      8         dybm.init_state()
----> 9         result_train = dybm.learn(in_seq, out_seq, get_result=True)
     10 
     11 

~/dybm-master/src/pydybm/time_series/time_series_model.py in learn(self, in_generator, out_generator, get_result)
    140                 predictions.append(prediction)
    141                 actuals.append(out_pattern)
--> 142             self.learn_one_step(out_pattern)
    143             self._update_state(in_pattern)
    144         return {"prediction": predictions, "actual": actuals}

~/dybm-master/src/pydybm/time_series/time_series_model.py in learn_one_step(self, out_pattern)
     85             pattern whose log likelihood is to be increased
     86         """
---> 87         delta = self._get_delta(out_pattern)
     88         if delta is not None:
     89             self._update_parameters(delta)

~/dybm-master/src/pydybm/time_series/dybm.py in _get_delta(self, out_pattern, expected)
    912 
    913     def _get_delta(self, out_pattern, expected=None):
--> 914         return ComplexDyBM._get_delta(self, [out_pattern])
    915 
    916     def get_predictions(self, in_seq):

~/dybm-master/src/pydybm/time_series/dybm.py in _get_delta(self, out_patterns, expected, weightLLs)
    653         deltas = list()
    654         for layer in self.layers:
--> 655             d = layer._get_delta(out_patterns, expected, weightLLs)
    656             deltas.append(d)
    657         return deltas

~/dybm-master/src/pydybm/time_series/vector_regression.py in _get_delta(self, out_patterns, expecteds, weightLLs)
    568         return [layer._get_delta(out_pattern, expected, weightLL)
    569                 for (layer, out_pattern, expected, weightLL)
--> 570                 in zip(self.layers, out_patterns, expecteds, weightLLs)]
    571 
    572     def _get_gradient(self, out_patterns, expecteds=None):

~/dybm-master/src/pydybm/time_series/vector_regression.py in <listcomp>(.0)
    567 
    568         return [layer._get_delta(out_pattern, expected, weightLL)
--> 569                 for (layer, out_pattern, expected, weightLL)
    570                 in zip(self.layers, out_patterns, expecteds, weightLLs)]
    571 

~/dybm-master/src/pydybm/time_series/vector_regression.py in _get_delta(self, out_pattern, expected, weightLL)
    180         """
    181         assert out_pattern.shape == (self.out_dim,), \
--> 182             "out_pattern must have shape (out_dim,)"
    183         if expected is not None:
    184             assert expected.shape == (self.out_dim,), \

AssertionError: out_pattern must have shape (out_dim,)

---------------------------------------------------------------------------

I've also 'corrected' the problem and would like to know if this is the intended behavior.

The problem resides in https://github.com/ibm-research-tokyo/dybm/blob/7895f4b8d1207e5af9fb81273a62e2b805bf0b12/src/pydybm/time_series/dybm.py#L872-L875

Since the out_dim value is never passed into this constructor, the value is rewritten and it will equal the in_dim
https://github.com/ibm-research-tokyo/dybm/blob/7895f4b8d1207e5af9fb81273a62e2b805bf0b12/src/pydybm/time_series/dybm.py#L574-L575

Therefore, to fix this I passed the out_dim value on line 873 above
`[in_dim], [out_dim], SGD=SGD, L1=L1, L2=L2,`


I would like to add that I did not check if other models, besides the BinaryDyBM, also have this problem.

Regards,
Alexandre Abreu

--- a/src/pydybm/time_series/dybm.py
+++ b/src/pydybm/time_series/dybm.py
@@ -869,8 +869,13 @@ class BinaryDyBM(ComplexDyBM):
         self.len_fifo = delay - 1
         self.in_dim = in_dim
         self.out_dim = out_dim
+        if out_dim is None:
+            out_dims = None
+        else:
+            out_dims = [out_dim]
+
         ComplexDyBM.__init__(self, [delay], [decay_rates], ["sigmoid"],
-                             [in_dim], SGD=SGD, L1=L1, L2=L2,
+                             [in_dim], out_dims=out_dims, SGD=SGD, L1=L1, L2=L2,
                              use_biases=[[True]], sigma=sigma,
                              insert_to_etrace=insert_to_etrace)
 

